gunrock/app/kcore/kcore_enactor.cuh

```diff
@@ -0,0 +1,578 @@
+                            const SizeT &input_pos, SizeT &output_pos) -> bool {
+
+        if (PRINT_DEBUG(!quiet, src, k) || PRINT_DEBUG(!quiet, dest, k)) {
+          printf("advance: src %u, dest %u, edgeID %u, input item %u, input pos %u, output pos %u\n, degrees %u, kcores %u\n", src, dest, edge_id, input_item, input_pos, output_pos, degrees[src], k_cores[src]);
+        }
+
+        if (!checkVertexRemains(delete_bitmap, src)) {
+          if (PRINT_DEBUG(!quiet, src, k) || PRINT_DEBUG(!quiet, dest, k)) {
+            printf("vertex %u deleted\n", src);
+          }
+            return false;
+        }
+
+        if(degrees[src] > k) {
+          if (PRINT_DEBUG(!quiet, src, k) || PRINT_DEBUG(!quiet, dest, k)) {
+            printf("deg(vertex) = %u > k (%u) for %u.\n", degrees[src], k, src);
+          }
+          return false;
+        }
+
+        else {
+          if (PRINT_DEBUG(!quiet, src, k) || PRINT_DEBUG(!quiet, dest, k)) {
+            printf("deg(vertex) = %u < k (%u) for %u.\n", degrees[src], k, src);
+          }
+          k_cores[src] = k;
+          markToDelete(to_be_deleted_bitmap, src);
+	  if(!checkVertexRemains(delete_bitmap, dest)){
+	    return false;
+	  }
+          return true;
+        }
+
+      };
+
+      auto filter_op = [degrees, k_cores, k, quiet, delete_bitmap, this
+      ] __device__(const VertexT &src, VertexT &dest,
+                            const SizeT &edge_id, const VertexT &input_item,
+                            const SizeT &input_pos, SizeT &output_pos) -> bool {
+
+        if (PRINT_DEBUG(!quiet, src, k) || PRINT_DEBUG(!quiet, dest, k)) {
+          printf("filter: src %u, dest %u, edgeID %u, input item %u, input pos %u, output pos %u\n, degrees %u, kcores %u\n", src, dest, edge_id, input_item, input_pos, output_pos, degrees[dest], k_cores[dest]);
+        }
+
+        if (!checkVertexRemains(delete_bitmap, dest)) {
+          if (PRINT_DEBUG(!quiet, src, k) || PRINT_DEBUG(!quiet, dest, k)) {
+            printf("vertex %u has been removed\n", dest);
+          }
+          return false;
+        }
+
+        int oldDegrees = atomicSub(degrees + dest, 1);
+
+        if (oldDegrees != (k + 1)) {
+          if (PRINT_DEBUG(!quiet, src, k) || PRINT_DEBUG(!quiet, dest, k)) {
+            printf("vertex %u degree = %u\n", dest, (oldDegrees - 1));
+          }
+          return false;
+        }
+
+        else {
+          if (PRINT_DEBUG(!quiet, src, k) || PRINT_DEBUG(!quiet, dest, k)) {
+            printf("vertex %u degree = %u\n", dest, (oldDegrees - 1));
+          }
+          return true;
+        }
+
+      };
+
+    // --
+    // Run
+
+    //Fill frontier with all vertices in graph for next iteration
+    SizeT num_nodes = graph.nodes;
+    frontier.queue_length = num_nodes;
+    GUARD_CU(frontier.V_Q()->ForEach(
+      initial_frontier,
+      [] __host__ __device__(VertexT & v, VertexT & i) { v = i; 
+      }, num_nodes, util::DEVICE, 0));
+
+    while(true) {
```

Comment:
Why do you loop from within Core here? Core should get repeatedly called until StopCondition() == true, so your loop would be repeated too.

Reply:
I basically want nested while loops:
- The inner loop starts with a frontier of all vertices, and ~removes vertices with degree <= k, decrementing the neighbors' degrees, repeating until there are no more vertices with degree <= k. 
- Then the outer loop checks if there are any vertices remaining in the graph, resets the frontier, and increments k. 

Now that you mention it, I suppose I should be able do this with if statements in the stop condition? Though one problem with that could be a potential slow down, because it appears that the stop condition gets called about 5 times for each iteration of Core.